Storage/src/StorageObject.php

```diff
@@ -676,34 +683,40 @@ public function downloadToFile($path, array $options = [])
      */
     public function downloadAsStream(array $options = [])
     {
-        $headers = $options['restOptions']['headers'] ?? [];
+        // This makes sure we honour the range headers specified by the user
+        $requestedBytes = $this->getRequestedBytes($options);
         $resultStream = Utils::streamFor(null);
-        $userSuppliedRangeHeaders = isset($headers['Range']);
-        $expectedSize = isset($options['size']) ? $options['size'] : $this->info()['size'];
+        $transcodedObj = false;
 
-        // add Range headers, which default to the whole object
-        // but only if the request doesn't already contain any range headers
-        if (!$userSuppliedRangeHeaders) {
-            $headers['Range'] = sprintf("bytes=%s-", $resultStream->getSize());
-            $options = array_merge($options, ['restOptions'=>['headers'=>$headers]]);
+        // We try to deduce if the object is a transcoded object when we receive the headers.
```

Comment:
This could maybe be split out into a function for clarity.

Reply:
So, the current way to get those headers are by the callback(anonymous func) I passed in the `restOptions.on_headers` option.

At the max I can extract the logic of the anonymous function in a prvt method, but I feel the context is only meaningful for the downloads and not at a class level.
WDYT?

Reply:
Sounds reasonable, I will leave it to your judgment